src/parallel.jl

```diff
@@ -0,0 +1,180 @@
+        end # chunk loop 
+
+        iter_time = toq();
+        total_time += iter_time
+        verbose ? @printf("[Iteration %-4d] residual: %10.3G | iteration runtime: %10.3f ms, (%10.3G s total)\n", i, residual[1], iter_time*1000.0, total_time) : nothing
+        verbose ? flush(STDOUT) : nothing
+        residual[1] < solver.belres ? break : nothing
+    end # main iteration loop
+    return ValueIterationPolicy(mdp, convert(Array{Float64, 2}, qmat), convert(Array{Float64, 1} , util), convert(Array{Int64, 1}, pol))
+end
+
+# updates shared utility and Q-Matrix using gauss-seidel value iteration (asynchronous)
+function solve_chunk(mdp::M, 
+                    states::SharedArray{S, 1}, 
+                    util::SharedArray{Float64, 1}, 
+                    pol::SharedArray{Int64, 1}, 
+                    qmat::SharedArray{Float64, 2}, 
+                    include_Q::Bool,
+                    residual::SharedArray{Float64, 1},
+                    state_indices::Tuple{Int64, Int64}
+                    ) where {M <: Union{MDP, POMDP}, S}
+
+    discount_factor = discount(mdp)
+    for istate=state_indices[1]:state_indices[2]
+        s = states[istate]
+        sub_aspace = actions(mdp, s)
+        if isterminal(mdp, s)
+            util[istate] = 0.0
+            pol[istate] = 1
+        else
+            old_util = util[istate] # for residual
+            max_util = -Inf
+            for a in iterator(sub_aspace)
+                iaction = action_index(mdp, a)
+                dist = transition(mdp, s, a) # creates distribution over neighbors
+                u = 0.0
+                for (sp, p) in weighted_iterator(dist)
+                    p == 0.0 ? continue : nothing # skip if zero prob
+                    r = reward(mdp, s, a, sp)
+                    isp = state_index(mdp, sp)
+                    u += p * (r + discount_factor * util[isp])
+                end
+                new_util = u
+                if new_util > max_util
+                    max_util = new_util
+                    pol[istate] = iaction
+                end
+                include_Q ? (qmat[istate, iaction] = new_util) : nothing
+            end # action
+            util[istate] = max_util
+            diff = abs(max_util - old_util)
+            diff > residual[1] ? (residual[1] = diff) : nothing
+            # update the value array
+        end
+    end # state loop
+    return 
+end
+
+# returns an array of start and end indices for each chunk
+function chunk_ordering(n_procs::Int64, order::Vector{Tuple{Int64, Int64}})
+    n_chunks = length(order)
+    # start and end indices
+    chunks = Vector{Vector{Tuple{Int64, Int64}}}(n_chunks) 
+    for i = 1:n_chunks
+        co = order[i]
+        start_idx = co[1]
+        end_idx = co[2]
+        n_states_per_chunk = end_idx - start_idx
+        # divide the work among the processors
+        stride = div(n_states_per_chunk, (n_procs-1))
+        temp = Vector{Tuple{Int64, Int64}}()

```

Comment:
We already know the length of `temp = n_procs-1`. Why not allocate accordingly and avoid `push!`?